sim4rec/response/nn_response.py

```diff
@@ -0,0 +1,204 @@
+            hist_data = spark.createDataFrame([], schema=SIM_LOG_SCHEMA)
+        # filter users whom we don't need
+        hist_data = hist_data.join(new_recs, on="user_idx", how="inner").select(
+            hist_data["*"]
+        )
+
+        # read the updated simulator log
+        simlog = spark.read.schema(SIM_LOG_SCHEMA).parquet(self.log_dir)
+        if not simlog:
+            print("Warning: the simulator log is empty")
+            simlog = spark.createDataFrame([], schema=SIM_LOG_SCHEMA)
+        # filter users whom we don't need
+        simlog = simlog.join(new_recs, on="user_idx", how="inner").select(simlog["*"])
+
+        NEW_ITER_NO = 9999999
+
+        # since all the historical records are older than simulated by design,
+        # and new slates are newer than simulated, i can simply concat it
+        combined_data = hist_data.unionByName(simlog).unionByName(
+            new_recs.withColumn("response_proba", sf.lit(0.0))
+            .withColumn("response", sf.lit(0.0))
+            .withColumn(
+                "__iter",
+                sf.lit(
+                    NEW_ITER_NO
+                ),  # this is just a large number, TODO: add correct "__iter" field to sim4rec.sample_responses to avoid this constants
+            )
+        )
+
+        # not very optimal way, it makes one worker to
+        # operate with one user, discarding batched computations.
+        # TODO: add batch_id column and use one worker ?
+        groupping_column = "user_idx"
+        result_df = combined_data.groupby(groupping_column).applyInPandas(
+            predict_udf, SIM_LOG_SCHEMA
+        )
+        filtered_df = result_df.filter(sf.col("__iter") == NEW_ITER_NO)
+        return filtered_df.select(new_recs.columns + [self.outputCol])
+
+
+class NNResponseEstimator(ActionModelEstimator):
+    def __init__(
+        self,
+        log_dir: str,
+        model_name: str,
+        hist_data_dir=None,
+        val_data_dir=None,
+        outputCol: str = "response_proba",
+        **kwargs,
+    ):
+        """
+        :param log_dir: The directory containing simulation logs.
+        :param model_name: Backbone model name.
+        :param hist_data_dir: (Optional) Spark DataFrame with historical data.
+        :param val_data_dir: (Optional) Spark DataFrame with validation data.
+                            TODO: split automatically.
+        :param outputCol: Output column for MLLib pipeline.
+
+        """
+        self.fit_params = kwargs
+        self.outputCol = outputCol
+
+        # sim log is not loaded immideately, because
+        # it can be not created when the response model is initialized
+        self.log_dir = log_dir
+        self.hist_data_dir = hist_data_dir
+        self.val_data_dir = val_data_dir
+
+        # create new model
+        self.item_indexer = self.user_indexer = None
+        self.model_name = model_name
+        self.backbone_response_model = None
+
+    def _fit(self, train_data):
+        """
```

Comment:
pls describe the dataframe format here and for transform. what should be included to properly condvert dataframe to the RecommendationData

Reply:
It is the same exactly, as the simulator logs format. Please give me advice, where I can obtain it's description.